ampd/src/broadcaster_v2/msg_queue.rs

```diff
@@ -0,0 +1,545 @@
+        deadline: time::Sleep,
+        queue: Queue,
+        duration: time::Duration,
+    }
+}
+
+impl MsgQueue {
+    pub fn new_msg_queue_and_client<T>(
+        broadcaster: broadcaster::Broadcaster<T>,
+        msg_cap: usize,
+        gas_cap: Gas,
+        duration: time::Duration,
+    ) -> Result<(impl Stream<Item = Vec<QueueMsg>>, MsgQueueClient<T>)>
+    where
+        T: cosmos::CosmosClient,
+    {
+        let (tx, rx) = mpsc::channel(msg_cap);
+
+        Ok((
+            Box::pin(MsgQueue {
+                stream: ReceiverStream::new(rx).fuse(),
+                deadline: time::sleep(duration),
+                queue: Queue::new(gas_cap),
+                duration,
+            }),
+            MsgQueueClient { broadcaster, tx },
+        ))
+    }
+}
+
+impl Stream for MsgQueue {
+    type Item = Vec<QueueMsg>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut me = self.as_mut().project();
+
+        loop {
+            match me.stream.as_mut().poll_next(cx) {
+                Poll::Pending => break,
+                Poll::Ready(Some(msg)) => {
+                    if me.queue.is_empty() {
+                        me.deadline.set(time::sleep(*me.duration));
+                    }
+
+                    if let Some(msgs) = me.queue.push(msg) {
+                        return Poll::Ready(Some(msgs));
+                    }
+                }
+                Poll::Ready(None) => {
+                    // returning Some here is only correct because we fuse the inner stream.
+                    let last = if me.queue.is_empty() {
+                        None
+                    } else {
+                        Some(me.queue.pop_all())
+                    };
+
+                    return Poll::Ready(last);
+                }
+            }
+        }
+
+        if me.queue.is_empty() {
+            return Poll::Pending;
+        }
+
+        match me.deadline.poll(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(()) => Poll::Ready(Some(me.queue.pop_all())),
+        }
+    }
+}
+
+struct Queue {
+    msgs: Vec<QueueMsg>,
+    gas_cost: Gas,
+    gas_cap: Gas,
+}
+
+impl Queue {
+    pub fn new(gas_cap: Gas) -> Self {
+        Queue {
+            msgs: vec![],
+            gas_cost: Gas::default(),
+            gas_cap,
+        }
+    }
+
+    pub fn push(&mut self, msg: QueueMsg) -> Option<Vec<QueueMsg>> {
+        match self.gas_cost.checked_add(msg.gas) {
+            None => {
+                let results = self.pop_all();
+
+                self.gas_cost = msg.gas;
+                self.msgs.push(msg);
+
+                Some(results)
+            }
+            Some(gas_cost) => {
+                self.gas_cost = gas_cost;
+                self.msgs.push(msg);
+
+                if self.gas_cost >= self.gas_cap {
+                    Some(self.pop_all())
+                } else {
+                    None
+                }
```

Comment:
I believe you need to compare gas_cost and gas_cap before pushing. We don't want to create a batch that's over capacity

Reply:
That's not true. If the estimate of one single msg already exceeds the cap, we still have to send it out without failing. The promise here is simply that, one cap is exceeded, the batch will be cut.

Reply:
Maybe `gas_threshold` is better name?

Reply:
Hmm, that can lead to message batches being too large to send though